include/dlaf/cuda/executor.h

```diff
@@ -0,0 +1,220 @@
+    // current device on the host thread. That could be the case if a previous task executing on the same
+    // host thread set a different device. [1]
+    //
+    // [1]: CUDA Programming Guide, section 3.2.3.6, Stream and Event Behavior
+    cudaError_t err = cudaEventQuery(event);
+
+    // Note: Despite the name, `cudaErrorNotReady` is not considered an error. [1]
+    //
+    // [1]: CUDA Runtime API, section 5.33 Data types used by CUDA Runtime
+    if (err == cudaErrorNotReady) {
+      return true;
+    }
+    else if (err == cudaSuccess) {
+      return false;
+    }
+    DLAF_CUDA_CALL(err);  // An error occured, report and terminate.
+    return false;         // Unreachable!
+  });
+  DLAF_CUDA_CALL(cudaEventDestroy(event));
+}
+
+}
+
+/// An executor for CUBLAS functions.
+///
+/// Design rationale:
+///
+/// There are a few alternative HPX/CUDA implementation options:
+///
+/// 1) The approach taken here inserts an event just after a CUBLAS function is called and uses it to
+///    track when the function has finished. Each task calls `hpx::util::yield_while()` and
+///    `cudaEventQuery()` to infrom HPX's scheduler when it's is ready.
+///
+/// 2) The current implementation in HPX inserts a callback function on a stream using
+///    `cudaStreamAddCallback()`. The callback registers the OS thread on which the stream is running
+///    with HPX and sets the data associated with a custom future type to indicate that a preceding CUDA
+///    function has completed.
+///
+///   There are a few issues with this approach:
+///
+///    - `cudaStreamAddCallback()` is deprecated in recent CUDA versions and is now superseded by
+///      `cudaLaunchHostFunc()`.
+///    - HPX throws an error if an external thread is registered multiple times, which happens as a
+///      result of the callback being called multiple times. To make this work, the error is currently
+///      ignored, which seems a little hacky.
+///    - The design is based on `targets` instead of `executors`.
+///
+/// 3) [NOT IMPLEMENTED] This approach is similar to how MPI Futures are implemented. `cudaEvent_t` are
+///    stored in a global array and polled for completion whenever a task yields or completes on a pool
+///    that has the polling function registered.
+///
+/// It is unclear at the moment which approach is the best. According to John 1) and 3) ought to have
+/// very similar performance. The two approaches are also available to MPI, performance comparisons
+/// there did not produce a clear winner which appears to confirm John's statement.
+///
+/// Most CUDA functions are thread safe, that includes Streams and CUBLAS handles. The executor accepts a
+/// `pool_executor` which specifies the host threads on which the CUBLAS calls are scheduled. There is a
+/// price to pay for potentially sharing the CUBLAS handles between multiple threads:
+///
+///  - `cudaSetDevice()` needs to precede the CUBLAS call to make it's handle valid on that thread
```

Comment:
That's not a problem and anyway always needed in hpx since the device set is thread local.

Reply:
With the oversubscription approach, we will be able to call `cudaSetDevice()` just once for the dedicate OS thread, instead of calling it every time a kernel is launched (like now). That's the difference.

Reply:
I would not consider a problem anyway:
cudaSetDevice() "should be considered a very low overhead call."
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g159587909ffa0791bbe4b40187a4c6bb

Reply:
This is similar to my concern with `cublasSetStream()`, I am aware that the docs consider it a very low overhead call, but I haven't benchmarked it and I'd expect at least an atomic or a mutex involved.

Reply:
cudaSetDevice() is thread local, why should a mutex or an atomic operation be needed?

Reply:
https://github.com/illuhad/hipSYCL/issues/87

Reply:
They have 7-10 calls per kernel launch, in our case it will be just once per kernel.
Moreover one of the developer stated:
> While that many calls to cudaSetDevice are certainly a little bizarre, I actually don't think they are related to your observed kernel launch latency."